pkg/southbound/synchronizer/factory.go

```diff
@@ -68,9 +70,54 @@ func Factory(topoChannel <-chan *topodevice.ListResponse, opStateChan chan<- eve
 					device:                    device,
 					target:                    newTargetFn(),
 				}
-				synchronizers[device.ID] = synchronizer
-				go synchronizer.connect()
+				connectionMonitors[device.ID] = connMon
+				go connMon.connect()
+			} else if ok && topoEvent.Type == topodevice.ListResponse_UPDATED {
+				changed := false
+				if connMon.device.Address != topoEvent.Device.Address {
+					oldAddress := connMon.device.Address
+					connMon.device.Address = topoEvent.Device.Address
+					changed = true
+					log.Infof("Topo device %s is being UPDATED - waiting to complete", device.ID)
+					connMon.close()
+					// TODO Change grpc.DialContext() used to non blocking so that we can
+					//  close the connection right away See https://github.com/onosproject/onos-config/issues/981
+					waitTime := *connMon.device.GetTimeout() //Use the old timeout in case it has changed
+					if maxBackoffTime > waitTime {
+						waitTime = maxBackoffTime
+					}
+					time.Sleep(waitTime + time.Millisecond*20) // close might not take effect until timeout
+					connMon.reopen()
+					go connMon.connect()
+					log.Infof("Topo device %s UPDATED address %s -> %s ", device.ID, oldAddress, topoEvent.Device.Address)
+				}
+				if connMon.device.Timeout.String() != topoEvent.Device.Timeout.String() {
+					connMon.mu.Lock()
+					oldTimeout := connMon.device.Timeout
+					connMon.device.Timeout = topoEvent.Device.Timeout
+					changed = true
+					connMon.mu.Unlock()
+					log.Infof("Topo device %s UPDATED timeout %s -> %s ", device.ID, oldTimeout, topoEvent.Device.Timeout)
+				}
+				if !changed {
+					log.Infof("Topo device %s UPDATE not supported %v", device.ID, device)
+				}
+			} else if ok && topoEvent.Type == topodevice.ListResponse_REMOVED {
+				log.Infof("Topo device %s is being REMOVED - waiting to complete", device.ID)
+				delete(connectionMonitors, device.ID)
+				connMon.close()
+				// TODO Change grpc.DialContext() used to non blocking so that we can
+				//  close the connection right away See https://github.com/onosproject/onos-config/issues/981
+				waitTime := *connMon.device.GetTimeout() //Use the old timeout in case it has changed
+				if maxBackoffTime > waitTime {
+					waitTime = maxBackoffTime
+				}
+				time.Sleep(waitTime + time.Millisecond*20) // close might not take effect until timeout
+				log.Infof("Topo device %s REMOVED", device.ID)
+			} else {
+				log.Warnf("Unhandled event from topo service %v", topoEvent)
 			}
+			commMonitorMu.Unlock()
```

Comment:
maybe we need to discuss this. but we are locking the whole map for the entire duration of the connection process. Is there a way we can only use the the read/write during the intiial reat at line 58  unlocking it immediately then do the operations and then lock, unlock before the update ? because in this way the whole operations are not locked but the r/w is.